bin/nrnivmodl.in

```diff
@@ -64,84 +67,178 @@ if [ "$1" = "-loadflags" ] ; then
-			set -e
-		else
-			files="$files $i"
-		fi
-	done
+  for i in "$@" ; do
+    i_hide=`hide_spaces "$i"`
+    if test -d "$i" ; then
+      set +e
+      if test "$UsingCMake" = "yes" ; then
+        for j in $(cd "$i"; unset LC_ALL; ls *.mod) ; do
+          files="$files $i_hide/$j"
+        done
+      else
+        files="$files $(unset LC_ALL; ls $i/*.mod)"
+        incs="$incs $(unset LC_ALL; ls $i/*.inc)"
+      fi
+      set -e
+    else
+      if test "$UsingCMake" = "yes" ; then
+        if test -e "$i" -o -e "$i.mod" ; then
+          files="$files $i_hide"
+        else
+          echo "Arg Error: \"$i\" is not a folder or mod file name or prefix"
+
+          echo ""
+          echo -n "  Mod file, folder args:"
+          for j in "$@" ; do
+            echo -n " \"$j\""
+          done
+          echo ""
+
+          exit 1
+        fi
+      else
+        files="$files $i"
+      fi
+    fi
+  done
 else
-	# Unset LC_ALL for consistent mod order
-	files=$(unset LC_ALL; ls *.mod)
+  # Unset LC_ALL for consistent mod order
+  files=$(unset LC_ALL; ls *.mod)
 fi
+
 files=`echo "$files" | sed 's/^ *//'`
 
-mfiles=""
+base_names=""
 for i in $files ; do
-	mfiles="$mfiles `basename $i`"
+  base_names="$base_names `basename -s .mod $i`"
 done
-mfiles=`echo "$mfiles" | sed 's/^ *//'`
 
-echo $files
-echo $mfiles
+if test "$UsingCMake" = "yes" ; then
+  echo -n "Mod files:"
+  for i in $files ; do
+    base_name=`basename -s .mod "$i"`
+    dir_name=`dirname "$i"`
+    echo -n " \"`unhide_spaces $dir_name`/$base_name.mod\""
+  done
+  echo ""
+  echo ""
+fi
+
+if [ ! -d $MODSUBDIR ] ; then
+  echo "Creating $MODSUBDIR directory for .o files."
+  echo
+  mkdir $MODSUBDIR
+fi
 
 files=`echo "$files" | sed 's/\.mod//g'`
 
 if test "$files" = '*' ; then
-	files=""
+  files=""
 fi
 
 cd $MODSUBDIR
-mdir=`pwd`
-for i in $files; do
-	rm -f "`basename $i`".mod
-	ln -s ../$i.mod .
-done
-
-if test "$incs" != "" ; then
-	for i in $incs ; do
-		rm -f "`basename $i`"
-		ln -s ../$i .
-	done
+mdir="`pwd`"
+
+if test "$UsingCMake" = "yes" ; then
+  # construct file to be included by makefile to specify mod to c rule when
+  # executed in $MODSUBDIR (child folder of launch location folder)
+  MODMAKE=makemod2c_inc
+  echo "" > $MODMAKE
+  for i in $files ; do
+    case "$i" in
+      /*) f=$i;; # absolute, fine as is
+      *)  f=../$i;; # relative
+    esac
+    base_name="`basename $f`"
+    dir_name="`dirname $f`"
+    echo "\
+./$base_name.c: `escape_spaces "$f.mod"`
+	@printf \" -> \$(C_GREEN)NMODL\$(C_RESET) \$<\\\n\"
+	(cd \"`unhide_spaces $dir_name`\"; MODLUNIT=\$(NRNUNITS) \$(NOCMODL) $base_name.mod -o \"$mdir\")
+
+./$base_name.o: ./$base_name.c
+	@printf \" -> \$(C_GREEN)Compiling\$(C_RESET) \$<\\\n\"
+	\$(COMPILE) -I\"`unhide_spaces $dir_name`\" \$(INCLUDES) @CMAKE_CXX_COMPILE_OPTIONS_PIC@ -c \$< -o \$@
+" >> $MODMAKE
+  done
+  if test "$nrnivmodl_cfiles" != "" ; then
+    ifs="$IFS"
+    IFS=';'
```

Comment:
nitpick : didn't understand this part. could we add some comment saying what's the logic here? (might be helpful for new people)

Reply:
Added the comment
```
    # I find it very difficult to construct shell variables which are
    # space separated lists of items containing spaces. So, for a cmake build,
    # the nrnivmodl_cfiles variable is required to be a ';' separated list
    # of c files which may contain spaces.   
    # Save the Internal Field Separator and interpret variables as ';'
    # separated list of items (which may contain spaces). Restore after 
    # processing nrnivmodl_cfiles.
```